### PR TITLE
fix: `fastapi` auto-instrumentation by removing `fastapi-slim` support

### PR DESCRIPTION
**Pull request by smoke**
# Description

`fastapi-slim` support was introduced via https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2683 as a response to `fastapi` maintainers adding a no-dependencies package called `fastapi-slim`.

The changes over `opentelemetry-instrumentation-fastapi` have broken auto-instrumentation via `opentelemetry-instrument` as reproduced here https://github.com/smoke/fastapi-autoinstrumentation-fails
causing `Skipping instrumentation fastapi: DependencyConflict: requested: "fastapi-slim~=0.111.0" but found: "None"` (when used with OTEL_PYTHON_LOG_LEVEL=debug as https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1745 is not yet available)

The `fastapi-slim` itself is discontinued from maintainers as it caused issues with users of it.
Given it causes issues with opentelemetry auto-instrumentation as well, it is removed as most feasible option.

see https://github.com/fastapi/fastapi/pull/11522#issuecomment-2264639417 
see https://fastapi.tiangolo.com/release-notes/#01120

Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2683#issuecomment-2271199419

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- see https://github.com/smoke/fastapi-autoinstrumentation-fails

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

**Comment by smoke** [comment]
Checks fail because of side flacky test `instrumentation-flask-0 pypy-3.8 Ubuntu` https://github.com/open-telemetry/opentelemetry-python-contrib/actions/runs/10280305618/job/28447490916

File: tox.ini [hunk 1]
```diff
@@ -556,8 +556,7 @@ commands_pre =
   fastapi: pip install opentelemetry-semantic-conventions@{env:CORE_REPO}\#egg=opentelemetry-semantic-conventions&subdirectory=opentelemetry-semantic-conventions
   fastapi: pip install opentelemetry-sdk@{env:CORE_REPO}\#egg=opentelemetry-sdk&subdirectory=opentelemetry-sdk
   fastapi: pip install opentelemetry-test-utils@{env:CORE_REPO}\#egg=opentelemetry-test-utils&subdirectory=tests/opentelemetry-test-utils
-  fastapi-main: pip install -r {toxinidir}/instrumentation/opentelemetry-instrumentation-fastapi/test-requirements.txt
-  fastapi-slim: pip install -r {toxinidir}/instrumentation/opentelemetry-instrumentation-fastapi/test-requirements-slim.txt
+  fastapi: pip install -r {toxinidir}/instrumentation/opentelemetry-instrumentation-fastapi/test-requirements.txt
   lint-instrumentation-fastapi: pip install -r {toxinidir}/instrumentation/opentelemetry-instrumentation-fastapi/test-requirements.txt
```
Comment (emdneto) [inline tox.ini:560]:
```suggestion
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -17,14 +17,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Fixed
 
+- `opentelemetry-instrumentation-fastapi` fix `fastapi` auto-instrumentation by removing `fastapi-slim` support, `fastapi-slim` itself is discontinued from maintainers
+  ([???](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/???))
```
Comment (emdneto) [inline CHANGELOG.md:21]:
```suggestion
  ([2783](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2783))
```

**Review by tammy-baylis-swi — APPROVED** [review]
Thank you for this fix, including issue comments and how to reproduce. The pyproject.toml and helper function changes also fixed my own local setup.

**Review by emdneto — APPROVED** [review]
Tested pointing one of my workloads to this branch and it's working again.

**Pull request merged**